File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Utilities/AttributeResolutionDirectiveSet.cs
//-----------------------------------------------------------------------
// <copyright file="AttributeResolutionDirectiveSet.cs" company="Microsoft">
//      All rights reserved.
// </copyright>
//-----------------------------------------------------------------------

namespace Microsoft.CommonDataModel.ObjectModel.Utilities
{
    using System.Collections.Generic;
    using System.Linq;

    public class AttributeResolutionDirectiveSet
    {
        public ISet<string> Set { get; set; }
        public ISet<string> SetRemoved { get; set; }
        private string SortedTag;

        public AttributeResolutionDirectiveSet(ISet<string> set = null)
        {
            if (set != null)
            {
                this.Set = new HashSet<string>(set);
            }
        }

        public AttributeResolutionDirectiveSet Copy()
        {
            AttributeResolutionDirectiveSet result = new AttributeResolutionDirectiveSet();
            if (this.Set != null)
                result.Set = new HashSet<string>(this.Set);
            if (this.SetRemoved != null)
                result.SetRemoved = new HashSet<string>(this.SetRemoved);
            result.SortedTag = this.SortedTag;
            return result;
        }

        public bool Has(string directive)
        {
            if (this.Set != null)
            {
                return this.Set.Contains(directive);
            }
            return false;
        }

        public void Add(string directive)
        {
            if (this.Set == null)
                this.Set = new HashSet<string>();
            // once explicitly removed from a set, never put it back
            if (this.SetRemoved?.Contains(directive) == true)
                return;
            this.Set.Add(directive);
            this.SortedTag = null;
        }

        public void Delete(string directive)
        {
            if (this.SetRemoved == null)
                this.SetRemoved = new HashSet<string>();
            this.SetRemoved.Add(directive);
            if (this.Set != null && this.Set.Contains(directive))
                this.Set.Remove(directive);
            this.SortedTag = null;
        }

        public void Merge(AttributeResolutionDirectiveSet directives)
        {
            if (directives != null)
            {
                if (directives.SetRemoved != null)
                {
                    // copy over the removed list first
                    foreach (string d in directives?.SetRemoved)
                    {
                        this.Delete(d);
                    }
                }
                if (directives.Set != null) 
                {
                    foreach (string d in directives?.Set)
                    {
                        this.Set.Add(d);
                    }
                }
                this.SortedTag = null;
            }
        }

        public string GetTag()
        {
            if (string.IsNullOrEmpty(this.SortedTag))
            {
                if (this.Set != null && this.Set.Count > 0)
                {
                    this.SortedTag = "";
                    List<string> sorted = this.Set.ToList();
                    sorted.Sort();
                    sorted.ForEach(d =>
                    {
                        this.SortedTag += "-" + d;
                    });
                }
            }

            if (!string.IsNullOrEmpty(this.SortedTag))
                return this.SortedTag;
            return "";
        }
    }
}

File: objectModel/TypeScript/Utilities/AttributeResolutionDirectiveSet.ts

export class AttributeResolutionDirectiveSet {
    public set: Set<string>;
    public setRemoved: Set<string>;
    public sortedTag: string;
    constructor(set?: Set<string>) {
        if (set) {
            this.set = new Set<string>(set);
        }
    }
    public copy(): AttributeResolutionDirectiveSet {
        const result: AttributeResolutionDirectiveSet = new AttributeResolutionDirectiveSet();
        if (this.set) {
            result.set = new Set<string>(this.set);
        }
        if (this.setRemoved) {
            result.setRemoved = new Set<string>(this.setRemoved);
        }
        result.sortedTag = this.sortedTag;

        return result;
    }
    public has(directive: string): boolean {
        if (this.set) {
            return this.set.has(directive);
        }

        return false;
    }
    public add(directive: string): void {
        if (!this.set) {
            this.set = new Set<string>();
        }
        // once explicitly removed from a set, never put it back.
        if (this.setRemoved && this.setRemoved.has(directive)) {
            return;
        }
        this.set.add(directive);
        this.sortedTag = undefined;
    }

    public delete(directive: string): void {
        if (!this.setRemoved) {
            this.setRemoved = new Set<string>();
        }
        this.setRemoved.add(directive);
        if (this.set) {
            if (this.set.has(directive)) {
                this.set.delete(directive);
            }
        }
        this.sortedTag = undefined;
    }

    public merge(directives: AttributeResolutionDirectiveSet): void {
        if (directives) {
            if (directives.setRemoved) {
                // copy over the removed list first
                directives.setRemoved.forEach((d: string) => {
                    this.delete(d);
                });
            }
            if (directives.set) {
                directives.set.forEach((d: string) => {
                    this.add(d);
                });
            }
            this.sortedTag = undefined;
        }
    }
    public getTag(): string {
        if (this.sortedTag === undefined) {
            if (this.set && this.set.size) {
                this.sortedTag = '';
                const sorted: string[] = Array.from(this.set)
                    .sort();
                sorted.forEach((d: string) => {
                    this.sortedTag += `-${d}`;
                });
            }
        }

        if (this.sortedTag) {
            return this.sortedTag;
        }

        return '';
    }
}
